code/iaas/logic/src/main/java/io/cattle/platform/process/containerevent/ContainerEventCreate.java

```diff
@@ -61,76 +67,104 @@
     AccountDao accountDao;
 
     @Inject
+    InstanceDao instanceDao;
+
+    @Inject
     LockManager lockManager;
 
     @SuppressWarnings("unchecked")
     @Override
     public HandlerResult handle(ProcessState state, ProcessInstance process) {
-        if ( !MANAGE_NONRANCHER_CONTAINERS.get() ) {
+        if (!MANAGE_NONRANCHER_CONTAINERS.get()) {
             return null;
         }
 
-        ContainerEvent event = (ContainerEvent)state.getResource();
-        Map<String, Object> inspect = (Map<String, Object>)DataUtils.getFields(event).get(FIELD_DOCKER_INSPECT);
+        final ContainerEvent event = (ContainerEvent)state.getResource();
 
-        String rancherUuid = getRancherUuidLabel(inspect);
-        if ( rancherUuid != null ) {
-            return null;
-        }
-
-        scheduleInstance(event, inspect);
-
-        return null;
-    }
-
-    void scheduleInstance(final ContainerEvent event, final Map<String, Object> inspect) {
-        final Long accountId = event.getAccountId();
-        final String externalId = event.getExternalId();
-
-        lockManager.lock(new ContainerEventInstanceLock(accountId, externalId), new LockCallbackNoReturn() {
+        HandlerResult result = lockManager.lock(new ContainerEventInstanceLock(event.getAccountId(), event.getExternalId()), new LockCallback<HandlerResult>() {
             @Override
-            public void doWithLockNoResult() {
-                Instance instance = objectManager.findAny(Instance.class, INSTANCE.ACCOUNT_ID, accountId,
-                        INSTANCE.EXTERNAL_ID, externalId);
-                if ( instance == null ) {
-                    instance = objectManager.newRecord(Instance.class);
-                    instance.setKind(InstanceConstants.KIND_CONTAINER);
-                    instance.setAccountId(accountId);
-                    instance.setExternalId(externalId);
-                    instance.setNativeContainer(true);
-                    setName(inspect, instance);
-                    setNetwork(inspect, instance);
-                    setImage(inspect, instance);
-                    setHost(event, instance);
-                    instance = objectManager.create(instance);
+            public HandlerResult doWithLock() {
+                Map<String, Object> inspect = (Map<String, Object>)DataUtils.getFields(event).get(FIELD_DOCKER_INSPECT);
```

Comment:
@ibuildthecloud You suggested using CollectionUtils.toMap here, but the comment and my reply got hidden when I pushed my updates.  Re-including my reply:

Looked into making this change, but CollectionUtils.toMap(Object) returns an empty map when the input is null. The input in this case the docker inspect and while a a null value is valid (no inspect is available for destroy events), an empty non-null map is not and it causes NPEs further down. And I'd rather just check whether or not inspect is null than its children as well.